pkg/k8s/exposer.go

```diff
@@ -2,131 +2,114 @@
 package k8s
 
 import (
-	"context"
 	"fmt"
 	"strconv"
 	"strings"
 
 	"github.com/glothriel/wormhole/pkg/listeners"
 	"github.com/glothriel/wormhole/pkg/peers"
-	"github.com/sirupsen/logrus"
 	"go.uber.org/multierr"
-	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 )
 
-type k8sServiceExposer struct {
-	namespace    string
-	child        listeners.Exposer
-	ownSelectors map[string]string
+type clientProvider interface {
+	New() (*kubernetes.Clientset, error)
 }
 
-func (factory *k8sServiceExposer) Add(app peers.App) (peers.App, error) {
+type fromInClusterConfigClientProvider struct{}
+
+func (fromInClusterConfigClientProvider) New() (*kubernetes.Clientset, error) {
 	config, inClusterConfigErr := rest.InClusterConfig()
 	if inClusterConfigErr != nil {
-		return peers.App{}, inClusterConfigErr
+		return nil, inClusterConfigErr
 	}
-	clientset, clientSetErr := kubernetes.NewForConfig(config)
+	return kubernetes.NewForConfig(config)
+}
+
+type k8sResourceExposer struct {
+	namespace string
+	child     listeners.Exposer
+	selectors map[string]string
+
+	managedResources []managedK8sResource
+
+	clientProvider clientProvider
+}
+
+func (exp *k8sResourceExposer) Add(app peers.App) (peers.App, error) {
+	clientset, clientSetErr := exp.clientProvider.New()
 	if clientSetErr != nil {
 		return peers.App{}, clientSetErr
 	}
-	servicesClient := clientset.CoreV1().Services(factory.namespace)
-	addedApp, childFactoryErr := factory.child.Add(app)
+	addedApp, childFactoryErr := exp.child.Add(app)
```

Comment:
I'm not a fan of nesting exposer within exposer - shoudn't we have some other interface that will accept a list of `exposers` and work on them accordingly?

`wormHoleKeeper` ? ;)

Reply:
I'd say it is what it is :D I didn't want to do a refactor of half the codebase. Saying that we can discuss it. Decorator is my favorite design pattern (as you probably noticed) and I may have hammer/nail problem with it.